Refactor to use data-logging-endpoint

AllanKerr opened the pull request:
# Changes
Rather than getting the provision endpoint from `D2L.Logging.GetLogger` it is now retrieved from the `data-logging-endpoint`

mtjandra commented on src/logger.js:
```diff
@@ -0,0 +1,26 @@
+function logError(
+	loggingEndpoint
```
Doesn't do anything yet, just wanted to verifying testing works correctly.

I don't setup projects all that often so I appreciate all the feedback so far

dbatiste commented on src/logger.js:
```diff
@@ -0,0 +1,26 @@
+	);
+}
+
+export const Logger = {
```
I am curious what the API will/should look like for this.  For example, we _could_ do something like:
```
export const Logger {

   constructor(endPoint) {
      ...
   }

   logError(...) {
      return new Promise((resolve, reject) => {
         ...
      });
   }

};
```
And consumers would...
```
import { Logger } from '@brightspace-ui/logging';

const logger = new Logger('some-end-point', ...);
logger.logError(...);
```
In the above scenario, consumers that have different end-points would need to create a new instance of Logger.  Do we want consumers to pass all the config for each `logError` call or not?  

I am interested in how the consuming app is going to use this - i.e. where it gets created, and if there's a shared one, how does it get a reference to it.

Aside, the returned promise may or may not be useful to the consumer, but as an asynchronous function, I think it's worthwhile returning it so the consumer can use it if it wants.

Interested in any/all other ideas.  ;)

mtjandra replied:
> Do we want consumers to pass all the config for each logError call or not?

I was building off @dlockhart earlier Slack suggestion

> The ES6 module could be the basic API that’s highly configurable and where you pass in the endpoint, but then the mixin could add a bit of sugar such that the endpoint is automatically snagged of the page like how we do localization

That being said maybe calling it `Logger` is confusing, perhaps I should call it `LoggingMethods`? Not sure.

To your other point

> the returned promise may or may not be useful to the consumer, but as an asynchronous function, I think it's worthwhile returning it so the consumer can use it if it wants

I would consider it "bad" design if a consumer relied on logging to work correctly. I don't want to promote something like.

```
Promise.resolve().then(() => logger.logError(new Error("Some Error"))).then( doSomething() );
```

If logging fails I couldn't think of anything worthwhile the consumer would do. Maybe retry, but if we want retry, I think the logger should handle that and not the consumer.


> I am interested in how the consuming app is going to use this

For the time being I was thinking along the lines of

```
import { Logger } from '@brightspace-ui/logging';

...

try {
  somethingThatMightFail();
}
catch( e ) {
  Logger.logError(this.loggingEndpoint, e, { devMessage: 'More Context if needed' });
}

...

this._getDataFromLMS()
  .then(data => ... )
  .catch( function (e) { Logger.logError(e); someOtherFailureHandling() } )
  .then( () => continueIfRecoverable() );
```

dlockhart replied:
So in my mind I was thinking the logging config would be passed into the constructor and then each log method would not need it. The big bonus there is that the logger could then be passed around/down and consumers could use it without needing to know the endpoint or other configuration data.

Now how and when it gets constructed and where that config data comes from is still an interesting question, but I think we can defer that a little bit.

dlockhart replied:
Oh and I agree with Miles that the promise probably isn't useful and might be dangerous. I could see us queuing up a bunch of log messages before sending them to reduce requests, and in that case folks waiting for promises would be detrimental. This should be a "fire and forget" type situation.

dbatiste replied:
For the promise, I was thinking of a scenario where the request could get abandoned.  Perhaps we don't care, perhaps we do.  But with the promise available, there could be code that makes sure promises are settled and messages not lost.  Thinking not just errors.  

I understand the concern about people doing something detrimental, but that comes with the territory of all asynchronous functions.  It is difficult to know all the scenarios, and providing the info (promise) gives them the choice in the same way that making the function `async` would automatically return a promise.  That all being said, we can always add it if the need arises.

mtjandra replied:
Sounds like for the time being we will not return anything from `logError` method, and we should construct the logger with the logging endpoint.

I will update to construct with logging endpoint

mtjandra commented:
@dbatiste @dlockhart I've made the updates to specify the logging endpoint in the constructor. This is ready to be looked at again whenever you guys are available. In the next PR I plan on putting in the logic to actually log messages instead of something with no info.

Commit:
Refactor to use data-logging-endpoint

Rather than getting the provision endpoint from D2L.Logging.GetLogger
it is now retrieved from the data-logging-endpoint

AllanKerr commented on logging.js:
```diff
@@ -34,14 +36,18 @@ export class ServerLogger {
-			logger = await this._loggerPromise;
-			options = { ...options, mode: 'no-cors' };
-			window.fetch(logger.Endpoint, options);
+		try {
```
I've wrapped these in `try` because they make a bunch of HTTP calls which could fail for any number of reasons. If logging fails I don't want an uncaught error interrupting execution. In that case the logs will just be output to the console.

dlockhart commented on logging.js:
```diff
@@ -50,11 +56,28 @@ export class ServerLogger {
+	async _provisionLoggerEndpoint(reload) {
+		const htmlEle = document.getElementsByTagName('html')[0];
+		if (!htmlEle) {
+			throw new Error(`Failed to locate top-level HTML element for ${dataLoggingEndpointAttribute}`);
```
Could these result in infinite errors if there's a `window.onerror` that calls into this code?

AllanKerr replied:
`_provisionLoggerEndpoint` is only ever `await`'ed in a `try` block so I think we should be safe there.
